pkg/operators/ebpf/helpers.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+)
+
+type (
+	btfTypeValidator func(btf.Type, string) (bool, error)
+	btfPopulateFunc  func(btf.Type, string) error
+	prefixFunc       func(string) (string, bool)
+	populateEntry    struct {
+		prefixFunc   prefixFunc
+		validator    btfTypeValidator
+		populateFunc btfPopulateFunc
+	}
+)
+
+func hasPrefix(prefix string) prefixFunc {
+	return func(s string) (string, bool) {
+		return strings.TrimPrefix(s, prefix), strings.HasPrefix(s, prefix)
+	}
+}
+
+var (
+	onceRingbuf      sync.Once
+	ringbufAvailable bool
+)
+
+func isRingbufAvailable() bool {
```

Comment:
This is never used and the logic to fallback is missing. Running a gadget in a kernel without this support fails.

Something like the following fixed the issue for me:
```diff
$ git diff
diff --git a/pkg/operators/ebpf/tracer.go b/pkg/operators/ebpf/tracer.go
index 26c2740f..8f2d2ae8 100644
--- a/pkg/operators/ebpf/tracer.go
+++ b/pkg/operators/ebpf/tracer.go
@@ -31,6 +31,11 @@ import (
        metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
 )
 
+// Keep this aligned with include/gadget/buffer.h
+const (
+       gadgetHeapMapName = "gadget_heap"
+)
+
 type Tracer struct {
        metadatav1.Tracer
 
@@ -109,6 +114,45 @@ func (i *ebpfInstance) populateTracer(t btf.Type, varName string) error {
                return fmt.Errorf("populating struct %q for tracer %q: %w", btfStruct.Name, name, err)
        }
 
+       // TODO: This should only be done if the tracer was created with the macro.
+       if err := i.handleTracerMapDefinition(tracerMap, btfStruct); err != nil {
+               return fmt.Errorf("handling tracer map definition: %w", err)
+       }
+
+       return nil
+}
+
+func (t *ebpfInstance) handleTracerMapDefinition(tracerMap *ebpf.MapSpec, btfStruct *btf.Struct) error {
+       heapMap, heapMapPresent := t.collectionSpec.Maps[gadgetHeapMapName]
+
+       if !isRingbufAvailable() {
+               if tracerMap != nil {
+                       log.Debugf("Ring buffers are not available, defaulting to perf ones")
+
+                       tracerMap.Type = ebpf.PerfEventArray
+                       tracerMap.KeySize = 4
+                       tracerMap.ValueSize = 4
+
+                       if heapMapPresent {
+                               // Replace MAX_EVENT_SIZE by the actual event size.
+                               heapMap.ValueSize = btfStruct.Size
+                       }
+               }
+       } else {
+               // If we delete the gadget_heap map, the verifier will not load the code
+               // because it cannot find it despite the code using it being dead one:
+               // ...: create BPF collection: program ig_mount_x: missing map gadget_heap
+               // Rather than deleting it, we just set its size to 0 and use a hash to
+               // avoid having one struct per CPU, i.e. reducing as much as possible the
+               // memory footprint it uses.
+               if heapMapPresent {
+                       heapMap.Type = ebpf.Hash
+                       heapMap.ValueSize = 4
+
+                       t.collectionSpec.Maps[gadgetHeapMapName] = heapMap
+               }
+       }
+
        return nil
 }
```

But I it needs extra checks to avoid changing the definition of a map provided by the user, i.e. not created with the GADGET_TRACER macro.